src/history/hs.c

```diff
@@ -971,11 +985,83 @@ __wt_find_hs_upd(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE **upd
 }
 
 /*
- * __hs_delete_key --
+ * __wt_hs_delete_key --
  *     Delete an entire key's worth of data in the history store.
  */
+int
+__wt_hs_delete_key(WT_SESSION_IMPL *session, uint32_t btree_id, const WT_ITEM *key)
+{
+    WT_CURSOR *hs_cursor;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    WT_TIME_PAIR hs_start, hs_stop;
+    uint32_t hs_btree_id, session_flags;
+    int cmp, exact;
+    bool close_hs_cursor;
+
+    session_flags = session->flags;
+    close_hs_cursor = false;
+
+    /*
+     * If we already have a history store cursor then don't open a new one. We do this while writing
+     * updates to the history store so it's important that we don't close the cursor since the upper
+     * layer logic will still be expecting it to be open.
+     */
+    if (!F_ISSET(session, WT_SESSION_HS_CURSOR)) {
+        WT_RET(__wt_hs_cursor(session, &session_flags));
+        close_hs_cursor = true;
+    }
+    hs_cursor = session->hs_cursor;
+    /*
+     * In order to delete a key range, we need to be able to inspect all history store records
+     * regardless of their stop time pairs.
+     */
+    F_SET(session, WT_SESSION_IGNORE_HS_TOMBSTONE);
+retry:
+    hs_cursor->set_key(hs_cursor, btree_id, key, WT_TS_NONE, WT_TXN_NONE, WT_TS_NONE, WT_TXN_NONE);
+    ret = hs_cursor->search_near(hs_cursor, &exact);
+    /* Empty history store is fine. */
+    if (ret == WT_NOTFOUND)
+        goto done;
+    WT_ERR(ret);
+    if (exact < 0)
+        ret = hs_cursor->next(hs_cursor);
+    /* That means there is only one record in the history store and it doesn't belong to our key. */
+    if (ret == WT_NOTFOUND)
+        goto done;
+    WT_ERR(ret);
+    /* Bailing out here also means we have no history store records for our key. */
+    WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_start.timestamp,
+      &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
+    if (hs_btree_id != btree_id)
+        goto done;
+    WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
+    if (cmp != 0)
+        goto done;
+    ret = __hs_delete_key_from_pos(session, hs_cursor, btree_id, key);
+    if (ret == WT_RESTART)
+        goto retry;
```

Comment:
Could add diagnostic code here to ascertain that all entries from HS are removed.